### ISSUE TITLE
Give users on a domain the ability to join their colleagues when the company is already using Expensify

### ISSUE THREAD

**Issue by anmurali** [Expensify/App#48189]
**Problem**: One of the cool things about Expensify is that anyone can sign up without their company's permission or invitation. But this means when an employee signs up independently they are not aware their company is already using Expensify and so end up creating their own workspaces in error, causing a lot of confusion for themselves and their company

**Solution**: Mirror Classic when it comes to an employee that signs up with a private domain email of a company that already has one or more workspaces. It goes like this:

1. Sign-up with an email address on a private domain that has existing workspace owners
2. We ask for their first name/last name (Ice breaker)
3. Next, we show a step that:
 - Let’s them know people on [derp.com](http://derp.com/) are already using Expensify.
 - We’ve sent a magic code to validate their account so they can join them
 - An input for said magic code to validate the account
 - An option to skip and move on with onboarding as normal if they prefer
4. If they validate the account, the next step shown is a list of all the workspaces owned by members of their private domain.
 - They can only pick one to join
5. If they click “Join team” we do one of two things depending on if the workspace pre-approves join requests, or an admin needs to accept:
 - Pre-approval not required:
    - Add them to the workspace
    - Drop them in the LHN
    - populate their onboarding/ConciergeDM with the invited member set of tasks
 - Pre-approval required:
   - Drop them in the LHN to continue exploring the product
   - Send the actionable message in the #admins room to accept or decline (just like policy join links that require approval)
   - On the workspaces page, show the workspace row they’re pending to be accepted to join with the “[Requested]” badge (just like a policy join request that needs to be approved)
   - If they select New Workspace in the + menu or Workspace page, have a warning just to let them know we’re waiting on their colleagues to still accept the request to join
   - Once accepted, they’ll get an email as they’re added to the workspace and we populate their onboarding/ConciergeDM with the invited member set of tasks.

**Comment by dubielzyk-expensify** [comment]
Here's the overall flow:

![CleanShot 2024-08-29 at 14 46 54@2x](https://github.com/user-attachments/assets/3a333bda-0421-4218-b358-32311d18ae63)

Included workspace member count and owner after feedback in Slack.

cc @Expensify/design for thoughts and feelz

**Comment by shawnborton** [comment]
Looks solid to me!

**Comment by dannymcclain** [comment]
>Looks solid to me!

Same!

>How does this look on desktop?

Also curious 🙃 👀

**Comment by dubielzyk-expensify** [comment]
Here's the desktop mocks:

![image](https://github.com/user-attachments/assets/deb6644b-cb23-4c02-b24a-612dd88414bc)
![image](https://github.com/user-attachments/assets/0182c698-ce4b-42d9-b2de-5c60f34ee8f6)

Do you think it's better if I don't stretch the validation code to be full-width with the modal?
![image](https://github.com/user-attachments/assets/2d968aa7-2284-4162-b416-8b928b8bd6a6)

**Comment by trjExpensify** [comment]
I like the full width version, personally! ❤️

**Comment by dannymcclain** [comment]
Agree! Stretch it baby! Desktop mocks looking good. 🫶

**Comment by dubielzyk-expensify** [comment]
The jury has spoken. Final Figma files can be found here: https://www.figma.com/design/ir27daDKHaB1g4iZCT22K5/Onboarding-v2?node-id=7147-9183

**Comment by anmurali** [comment]
@trjExpensify - can this be built by the contributor or do we need BE code? Curious for a second set of eyes. Asking you as you wrote down the steps in the solution

**Comment by trjExpensify** [comment]
I think this is mostly FE, but we might need some BE work for the "Select a workspace" page to get that list of accessibleDomainPolicies - I'm not quite sure if we have that available to NewDot. 

@luacmartins & @mountiny we built the request to join stuff in simplified collect with the `Share` policyJoinLink a while back, so the #admins message to accept or decline (if pre-approval is required) should be in place and usable here already, right?

**Comment by trjExpensify** [comment]
Cool, so should be reusable to call that request to join from here when needed with `Ask to join`: 

<img width="296" alt="image" src="https://github.com/user-attachments/assets/682a62bb-c3c2-4250-a0c7-bd1fae94586f">

>but we might need some BE work for the "Select a workspace" page to get that list of accessibleDomainPolicies - I'm not quite sure if we have that available to NewDot

Do you have any idea if we'd have a list of accessibleDomainPolicies available in NewDot or if that's going to need BE work here?

**Comment by luacmartins** [comment]
>Do you have any idea if we'd have a list of accessibleDomainPolicies available in NewDot or if that's going to need BE work here?

AFAIK we don't have this data in NewDot yet, so we'd need to fetch it when we open that page.

**Comment by trjExpensify** [comment]
Kewl, so we do need BE here then yes?

**Comment by luacmartins** [comment]
yes

**Comment by mountiny** [comment]
Yeah, agreed we will need to fetch that; also, somehow, indicate which one is "Join now" and which one is "Ask to join". couple questions:
1. How do the "Join Now" and "Ask to join" changes when you click on them?
2. If user clicks skip for now, where will they access this modal/ list of workspaces to join with buttons later in the product?

**Comment by trjExpensify** [comment]
>How do the "Join Now" and "Ask to join" changes when you click on them?

Join now = workspace doesn't require approval to join 
Ask to join = workspace needs admin approval to join

That's the distinction between the two buttons, so I _assume_ that's something on the policy object we store for use with policy join links. 

As for what happens when you click them, that's outlined in the OP in step 5. Let me know if you have any qs on that! 

>If user clicks skip for now, where will they access this modal/ list of workspaces to join with buttons later in the product?

If we're mirroring the Inbox on OldDot, they don't. So maybe "Skip" is better than "Skip for now" which implies you can get back to it. We could look at something on the workspaces page with a list of available workspaces to join instead of the empty state when applicable, but I'm also fine to descope that and keep it to this one-time onboarding modal flow.

**Comment by anmurali** [comment]
>That's the distinction between the two buttons, so I assume that's something on the policy object we store for use with policy join links.

I have clarified the exact workspace setting in the original description!

**Comment by anmurali** [comment]
**Final mockuops** 

Mobile - Employee
<img width="645" alt="image" src="https://github.com/user-attachments/assets/8d355997-23f7-40bb-8039-c1a94530ef92">

Web - Employee
<img width="816" alt="image" src="https://github.com/user-attachments/assets/6542731b-7cb6-4211-bdcb-420074465042">

Actionable whisper in # admins (Exists)

![image](https://github.com/user-attachments/assets/60e06384-0e91-4bd8-b8fc-99ba25ee4733)

**Comment by marcaaron** [comment]
>populate their onboarding/ConciergeDM with the invited member set of tasks

Just for reference... what are these exactly? Is this built already or a new set of tasks we are showing? Is it the same set of tasks when you select "Get paid back by my employer"?

>An option to skip and move on with onboarding as normal if they prefer

In other words, they will land back on this screen?

![2024-09-09_17-11-28](https://github.com/user-attachments/assets/fb6b4160-0a25-4733-96a7-adbe80055931)

**Comment by marcaaron** [comment]
Looking into returning the accessible domain policies during `OpenApp` (I assume that should be fine) as a first step here.

I think we will just need one new command for joining the accessible policy - how about `JoinAccessiblePolicy`? 

And seems like we will be able to use the existing `RequestPolicyAccess` when the user is requesting.

**Comment by trjExpensify** [comment]
>I think we will just need one new command for joining the accessible policy - how about JoinAccessiblePolicy?

Sounds like a solid name for this to me. 👍

**Comment by anmurali** [comment]
> populate their onboarding/ConciergeDM with the invited member set of tasks
>- Just for reference... what are these exactly? Is this built already or a new set of tasks we are showing? Is it the same set of tasks when you select "Get paid back by my employer"?

@marcaaron - this is being built as part of Stage 3 onboarding ([1.b](https://docs.google.com/document/d/1EX1tKSkhffpqP4TXD7iKIPKWnU5U8ELv32j-QTtXH2g/edit#heading=h.xbz08btaavfp))

**Comment by anmurali** [comment]
>An option to skip and move on with onboarding as normal if they prefer
>- In other words, they will land back on this screen?

Yes

**Comment by marcaaron** [comment]
Working on the backend changes for this now.

**Comment by marcaaron** [comment]
Working on the backend changes for this now.

**Comment by marcaaron** [comment]
Re-adding the Help Wanted label. Seems like Melvin removed it. 

Still working on the backend changes for this. Unfortunately, it's a little bit tricky to implement since it essentially adds a new email validation flow for a new user who has signed in with an unvalidated login. When they enter their magic code we need to also create a passwordless login if I'm not mistaken. 

Could actually use a second opinion from @johnmlee101 or @NikkiWines on that whenever they get a chance.

**Comment by marcaaron** [comment]
Let's get some proposals here for the frontend? It may be easier to implement the backend changes with the frontend partially built.

**Comment by marcaaron** [comment]
@cretadn22 did you want to revise your proposal? @parasharrajat how can we get this one moving?

**Comment by marcaaron** [comment]
Still working through the BE changes in these two PRs:

- [ ] https://github.com/Expensify/Auth/pull/12437
- [ ] https://github.com/Expensify/Web-Expensify/pull/43617

**Comment by marcaaron** [comment]
Still working on the backend stuff, but it's all in review now. The changes required some consensus from engineering on how to build it out and turned out to be pretty involved.

I'm going OOO starting on Friday and would love to try and get the PRs merged before then so this will be in a potentially good state to hand off to someone else.

**Comment by marcaaron** [comment]
Back from OOO, but the backend PRs still in review here. This ticket required creating a new authentication related command and turned out to be a bit involved. Getting the accessible domain policies is a performance sensitive operation as well. But I think we are _nearly_ there with @danieldoglas's great assistance.

**Comment by anmurali** [comment]
That original job post is _closed_ for some reason. So I created a new one with that higher price - https://www.upwork.com/ab/applicants/1844368510556052644

**Comment by anmurali** [comment]
@allgandalf still open!

**Comment by marcaaron** [comment]
>we need to verify if their domain already exists. This task should be handled in the backend

@cretadn22 Hmm not too sure what you are asking when you say "domain already exists". This will be checked whenever the user has a non-public domain so we can use the `PUBLIC_DOMAINS` constant right?

As for API, we will have two commands: `JoinAccessiblePolicy` and `ValidateUserAndGetAccessiblePolicies` will be required to complete this task.

@allgandalf yes, we are still accepting proposals.

**Comment by bronwynbiro** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
Users who register with a private domain should be able to see a list of workspaces within their domain, and have the option to join them. 

### What is the root cause of that problem?
This is a new feature

### What changes do you think we should make in order to solve the problem?
My proposal for how to handle each step is in bold below.

**1. Check if the email belongs to an existing domain:** 
**- [if email is not public domain,](https://github.com/Expensify/App/blob/main/src/libs/LoginUtils.ts#L37) call new API `ValidateUserAndGetAccessiblePolicies` to retrieve list of workspaces. We need `workspaceName, automaticJoiningEnabled,workspaceMembeCount,ownerEmail` for each workspace.** 
2.  We ask for their first name/last name
3.  **If there are any workspaces returned from step 1, adjust [adaptOnboardingRouteState](https://github.com/Expensify/App/blob/476d112f016a914013515414727d7019ff1a8c38/src/libs/actions/Welcome/OnboardingFlow.ts#L61C11-L61C40)**  and we show a step that:

-   Let’s them know people on  [derp.com](http://derp.com/)  are already using Expensify.
	- **change the header to display the "People you may know are already here!..." text in the [SignInPage.tsx file](https://github.com/Expensify/App/blob/476d112f016a914013515414727d7019ff1a8c38/src/pages/signin/SignInPage.tsx#L247)  if applicable**
-   We’ve sent a magic code to validate their account so they can join them
-   An input for said magic code to validate the account
	**- These steps can reuse the ValidateCode and ValidateCodeAction components**
-   An option to skip and move on with onboarding as normal if they prefer
	- **use [adaptOnboardingRouteState](https://github.com/Expensify/App/blob/476d112f016a914013515414727d7019ff1a8c38/src/libs/actions/Welcome/OnboardingFlow.ts#L61C11-L61C40) to keep state consistent and allow to skip this new workflow** 

5.  If they validate the account, the next step shown is a list of all the workspaces owned by members of their private domain.
	**- Create new screen for this view, maybe called `onboardingAvailableWorkspaces`**
		- **We need `workspaceName, automaticJoiningEnabled,workspaceMemberCount,ownerEmail` for each workspace. We will call the new API `ValidateUserAndGetAccessiblePolicies` in step 1 to get this info.** 

-   The CTA is  `Ask to join`  or  `Join`  based on the workspace JSON's  `automaticJoiningEnabled`  parameter being set to  `false`  or  `true`  respectivey
-   They can only pick one workspace in the list
**-  Call new API: `JoinAccessiblePolicy`**

6.  If they click on the CTA we do one of two things depending on if the workspace pre-approves join requests (Join), or an admin needs to accept (Ask to join):

-   Pre-approval not required:
    -   Add them to the workspace
    -   Drop them in the LHN
    -   populate their onboarding/ConciergeDM with the invited member set of tasks
    **- call [addMembersToWorkspace](https://github.com/Expensify/App/blob/476d112f016a914013515414727d7019ff1a8c38/src/libs/actions/Policy/Member.ts#L608)**
    **- [completeOnboarding](https://github.com/Expensify/App/blob/0984f7d006dafde324fcf041c8a96eeaf36069f5/src/libs/actions/Report.ts#L3325) may also need to be called?**
-   Pre-approval required:
    -   Drop them in the LHN to continue exploring the product
    -   Send the actionable message in the #admins room to accept or decline (just like policy join links that require approval)
	    - **this functionality exists already**
	    - **call [`inviteToRoom`](https://github.com/Expensify/App/blob/0984f7d006dafde324fcf041c8a96eeaf36069f5/src/libs/actions/Report.ts#L2898) and make [API call](https://github.com/Expensify/App/blob/0984f7d006dafde324fcf041c8a96eeaf36069f5/src/libs/actions/Policy/Member.ts#L705)** 
    -   On the workspaces page, show the workspace row they’re pending to be accepted to join with the “[Requested]” badge (just like a policy join request that needs to be approved)
    **- this functionality already exists [here](https://github.com/Expensify/App/blob/476d112f016a914013515414727d7019ff1a8c38/src/pages/workspace/WorkspacesListRow.tsx#L62)** 
    -   If they select New Workspace in the + menu or Workspace page, have a warning just to let them know we’re waiting on their colleagues to still accept the request to join
    **- will check if this exists, otherwise can add a check if the user's membership request is pending**
    -   Once accepted, they’ll get an email as they’re added to the workspace and we populate their onboarding/ConciergeDM with the invited member set of tasks.
    **- this should exist already, will update the proposal if not**

I wanted to get a draft proposal out first to check if the overall idea seems reasonable. I'll come back to edit some of the unknowns towards the bottom in a couple hours. Let me know if there's anything else I should add/if anything is unclear. Thanks!

**Comment by marcaaron** [comment]
@Bronwyn.biro@gmail.com proposal LGTM 👍

**Comment by marcaaron** [comment]
BE PRs are deployed. I think we can pull out the stops here. Going to assign @bronwynbiro to get the ball rolling.

**Comment by bronwynbiro** [comment]
Contributor details
Your Expensify account email: Bronwyn.biro@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01d6ad9121672abba2

**Comment by bronwynbiro** [comment]
> Still working through the BE changes in these two PRs:
> 
> * [ ]  https://github.com/Expensify/Auth/pull/12437
> * [x]  https://github.com/Expensify/Web-Expensify/pull/43617

I can't view these repos, it possible to get access? Or the expected input/output for `JoinAccessiblePolicy` and `ValidateUserAndGetAccessiblePolicies`? Thanks :)

**Comment by marcaaron** [comment]
Please continue with the frontend changes, both BE tasks have been deployed.

**Comment by marcaaron** [comment]
`JoinAccessiblePolicy` takes these params:

`policyID: string`

`ValidateUserAndGetAccessiblePolicies` takes

`validateCode: string` (6 digits)

The main thing to be aware of is that the accessible policies will be returned to an state key (Onyx) `KEY_JOINABLE_POLICIES` which is a string of `'joinablePolicies'` in the local data and this should be visible in the response under `onyxData`. Feel free to flag me down in Slack if you have any questions.

**Comment by bronwynbiro** [comment]
Thanks @marcaaron! Working on this now.

Some initial questions:
- How should I be generating the magic code? I was using `ValidateCodeForm` but after entering the magic code from my email, I get an error that it does not match the expected code. I think this is because I am calling that from a new account, whereas it is intended for an already verified account..
- Should a default onboarding purpose be set when the user skips onboarding? Or should it be left incomplete?

**Comment by marcaaron** [comment]
@bronwynbiro do you have a PR raised that I can look at and test to see where the problem might be?

Using `ResendValidateCode` command should work.

**Comment by marcaaron** [comment]
>I think this is because I am calling that from a new account, whereas it is intended for an already verified account

Not quite. Asking for the `validateCode` is only necessary on accounts that are **not validated**. We need them to confirm their identity then we will show them the accessible policies.

**Comment by marcaaron** [comment]
>An option to skip and move on with onboarding as normal if they prefer

I think the intention here is to "skip" only the validate and join policy step. No other steps should be skippable unless they are already skippable (which I think is none IIRC).

**Comment by bronwynbiro** [comment]
Thanks, got the validation bug sorted. I'm now a bit unclear on how the BE calls should look. I created an action:

```
function validateUserAndGetAccessiblePolicies(validateCode: string) {
    const parameters: ValidateUserAndGetAccessiblePoliciesParams = {validateCode: validateCode};
    API.read(READ_COMMANDS.VALIDATE_LOGIN_AND_SHOW_WORKSPACES, parameters);
}

```

but it seems the Onyx object need to be passed in to `API.read` as well? Something like..
```
 const optimisticData: OnyxUpdate[] = [
        {
            onyxMethod: Onyx.METHOD.MERGE,
            key: ONYXKEYS.KEY_JOINABLE_POLICIES,
            value: [],
        },
    ];
  ```
but what should the key and value be? or is there a different method I should be using?

**Comment by marcaaron** [comment]
No need pass in an empty array. It will actually be an object keyed by `policyID` (but also no need provide). 

We are going to want to have this block the user and show some kind of loading state while we are loading the policies. So we will need to follow an example similar to [this here](https://github.com/Expensify/App/blob/286212cc4f23b94426dca0bc28a039f61ca5c693/src/libs/actions/BankAccounts.ts#L211C5-L251C7).

**Comment by bronwynbiro** [comment]
Thank you! I'm getting a 500 error when making the request, so I think I'm still missing something

```
return new Promise((resolve, reject) => {
        const parameters: ValidateUserAndGetAccessiblePoliciesParams = {validateCode: validateCode};

        const optimisticData: OnyxUpdate[] = [
            {
                onyxMethod: Onyx.METHOD.MERGE,
                key: ONYXKEYS.KEY_JOINABLE_POLICIES,
                value: {isLoading: true},
            },
        ];

        const successData: OnyxUpdate[] = [
            {
                onyxMethod: Onyx.METHOD.MERGE,
                key: ONYXKEYS.KEY_JOINABLE_POLICIES,
                value: {isLoading: false},
            },
        ];

        const failureData: OnyxUpdate[] = [
            {
                onyxMethod: Onyx.METHOD.MERGE,
                key: ONYXKEYS.KEY_JOINABLE_POLICIES,
                value: {isLoading: false},
            },
        ];
    API.makeRequestWithSideEffects(SIDE_EFFECT_REQUEST_COMMANDS.VALIDATE_LOGIN_AND_SHOW_WORKSPACES, parameters, {optimisticData, successData, failureData})
.then((response) => {
  ...
```

the response looks like this: 
`"jsonCode":500,"message":"500 Can't call this command on public domains","onyxData":[],...`

**Comment by marcaaron** [comment]
yes, because the API should only be used on emails with private domains:

>Solution: Mirror Classic when it comes to an employee that signs up with a **private domain email** of a company that already has one or more workspaces.

**Comment by bronwynbiro** [comment]
Ahh thanks, forgot I was using my own email to test.. 

Is there a test private domain account I could use? I only have a gmail account, but being able to use a private domain account would help with testing the code verification and getting the list of available policies

**Comment by marcaaron** [comment]
Sorry, I don't have anything like that to offer right now. Maybe @allgandalf can assist with the testing portion. What I can do is try to provide you with mock data that you can hardcode to work on the frontend.

**Comment by bronwynbiro** [comment]
Sounds good, yeah hardcoded mock data would be great! I also opened a draft PR [here](https://github.com/Expensify/App/pull/51102)

**Comment by marcaaron** [comment]
The data will be something like this (omitted the fields not relevant for this ticket):

```json
{
  "D35BAB0CFAC69F52": {
    "policyID": "D35BAB0CFAC69F52",
    "policyOwner": "bob@domain.com",
    "policyName": "Health Coverage",
    "employeeCount": 150,
    "hasPendingAccess": false,
    "automaticJoiningEnabled": true,
  },
  "E56CAB1DFBC78A13": {
    "policyOwner": "alice@domain.com",
    "policyID": "E56CAB1DFBC78A13",
    "policyName": "Travel Support",
    "employeeCount": 50,
    "hasPendingAccess": true,
    "automaticJoiningEnabled": false,
  },
  "F67DAC2EABC89B24": {
    "policyOwner": "eve@domain.com",
    "policyID": "F67DAC2EABC89B24",
    "policyName": "Tech Equipment",
    "employeeCount": 200,
    "hasPendingAccess": false,
    "automaticJoiningEnabled": false,
  }
}
```

**Comment by isabelastisser** [comment]
Not overdue.

**Comment by bronwynbiro** [comment]
Hey all, wanted to provide an update. I'm currently stuck on the final steps (step 6) of this PR and won't be able to continue working on it beyond the end of day Thursday, as I'll be away starting Friday. I would appreciate any guidance or support to help wrap this up before then!

**Comment by bronwynbiro** [comment]
Until dec 5 🙃

**Comment by marcaaron** [comment]
let's see if we can get it wrapped and if not maybe you can take over @allgandalf and we will get another C+?

**Comment by isabelastisser** [comment]
@rushatgabhane @allgandalf, any updates here?

**Comment by isabelastisser** [comment]
@rushatgabhane, sorry! I didn't mean to tag you here. Please ignore!

**Comment by allroundexperts** [comment]
Taking over the development.

**Comment by allroundexperts** [comment]
Moving forward from [here](https://github.com/Expensify/App/pull/51681)

**Comment by allroundexperts** [comment]
@marcaaron The data currently is like:
```
{
  "isLoading": false,
  "D35BAB0CFAC69F52": {
    "policyID": "D35BAB0CFAC69F52",
    "policyOwner": "bob@domain.com",
    "policyName": "Health Coverage",
    "employeeCount": 150,
    "hasPendingAccess": false,
    "automaticJoiningEnabled": true,
  },
  "E56CAB1DFBC78A13": {
    "policyOwner": "alice@domain.com",
    "policyID": "E56CAB1DFBC78A13",
    "policyName": "Travel Support",
    "employeeCount": 50,
    "hasPendingAccess": true,
    "automaticJoiningEnabled": false,
  },
  "F67DAC2EABC89B24": {
    "policyOwner": "eve@domain.com",
    "policyID": "F67DAC2EABC89B24",
    "policyName": "Tech Equipment",
    "employeeCount": 200,
    "hasPendingAccess": false,
    "automaticJoiningEnabled": false,
  }
}
```

Instead of this, I think it makes more sense to have it like:

```
{
  "isLoading": false,
  "policies": {
      "D35BAB0CFAC69F52": {
        "policyID": "D35BAB0CFAC69F52",
        "policyOwner": "bob@domain.com",
        "policyName": "Health Coverage",
        "employeeCount": 150,
        "hasPendingAccess": false,
        "automaticJoiningEnabled": true,
      },
      "E56CAB1DFBC78A13": {
        "policyOwner": "alice@domain.com",
        "policyID": "E56CAB1DFBC78A13",
        "policyName": "Travel Support",
        "employeeCount": 50,
        "hasPendingAccess": true,
        "automaticJoiningEnabled": false,
      },
      "F67DAC2EABC89B24": {
        "policyOwner": "eve@domain.com",
        "policyID": "F67DAC2EABC89B24",
        "policyName": "Tech Equipment",
        "employeeCount": 200,
        "hasPendingAccess": false,
        "automaticJoiningEnabled": false,
      }
   }
}
```

Let me know if this is possible / quickly achievable.

**Comment by marcaaron** [comment]
I think this is a blocking request so we don't need to have `isLoading` in the Onyx data at all. Let's implement it that way for now? Or another option is to use a separate Onyx key just for the loading. Either seem acceptable to me.

**Comment by allroundexperts** [comment]
> I think this is a blocking request so we don't need to have isLoading in the Onyx data at all. 

But right now, `isLoading` is returned from the backend.

> Let's implement it that way for now?

The issue here is that doing an `Object.values` on the data would also result in iterating over the `isLoading` attribute. This breaks my logic until I do a filter which seems a work around to me.

**Comment by marcaaron** [comment]
>But right now, isLoading is returned from the backend.

I feel like this can't possibly be true, but maybe you can describe what you are seeing in more detail. Feel free to DM me about this.

**Comment by allroundexperts** [comment]
PR ready for review https://github.com/Expensify/App/pull/51681

**Comment by marcaaron** [comment]
@allroundexperts Gonna make this a Daily to hopefully help move it along. Feels like we are close!

**Comment by allroundexperts** [comment]
@marcaaron Can I get access to Figma for this? I'm a little confused on what the actual flow is. In the PR, the personal details step is the first step, while in the design screenshot that Anu posted has it as a last step.

**Comment by shawnborton** [comment]
@allroundexperts I think this is the flow we're expecting:
![image](https://github.com/user-attachments/assets/19ac3f54-1e64-4361-bbff-47b70e6de81d)


Does that look right @dubielzyk-expensify ?

**Comment by marcaaron** [comment]
@allroundexperts AFAIK yes that is correct. I'm not sure what you are specifically confused about though. Can you clarify what is meant by "personal details step" just to be sure we are understanding each other here?

**Comment by dubielzyk-expensify** [comment]
The intent step is actually first in most cases, then name. But in this instance I think we want to skip the intent step? Cause I think we have some weird follow-on things if you choose "Manage my team" etc which we don't want in this particular flow.

So in that instance what @shawnborton has posted above seems right to me. I guess like mentioned in the other GH, we probably wanna not have the back arrow on "Select a workspace" to end up in a double validation state. Unless we can neatly ignore that part when going back.

**Comment by marcaaron** [comment]
I believe some things got lost in translation here and it was not clear to us that these steps would be in scope for this issue:

>On the workspaces page, show the workspace row they’re pending to be accepted to join with the “[Requested]” badge (just like a policy join request that needs to be approved)

>If they select New Workspace in the + menu or Workspace page, have a warning just to let them know we’re waiting on their colleagues to still accept the request to join

Just a heads up that we are going to proceed without that for now since the PR is close to merging. But also, did we have any mocks for this? Maybe Design can weigh in here. Also, unclear what kind of warning we will show if they attempt to create a workspace. Perhaps these can be broken into some follow up tickets? 

cc @shawnborton @dubielzyk-expensify @dannymcclain

**Comment by dubielzyk-expensify** [comment]
> On the workspaces page, show the workspace row they’re pending to be accepted to join with the “[Requested]” badge (just like a policy join request that needs to be approved)

Here's a take:
![image](https://github.com/user-attachments/assets/fc06cf55-5048-4846-90b0-c2c9946d7b2a)

Also worth noting that [we're adding a badge for default/primary workspace](https://github.com/Expensify/App/issues/52373) so it'll be a very similar treatment. A PR is in progress there.

> If they select New Workspace in the + menu or Workspace page, have a warning just to let them know we’re waiting on their colleagues to still accept the request to join

I'm not 100% sure we should do this if we're adding the pending label personally. Keen to hear @Expensify/design thoughts.

If we did do a warning, I recon we'd use a normal confirmation dialog (just with better copy):
![image](https://github.com/user-attachments/assets/0d2c9562-02c0-499a-ba63-2b69f1b1e8f9)

**Comment by shawnborton** [comment]
> If we did do a warning, I recon we'd use a normal confirmation dialog (just with better copy):

Agree that if we do decide to do this, that mockup is what I would expect. I don't feel too strongly really, I can see where it might be beneficial in this case to make it slightly harder to make the new workspace. But keep in mind we're also going to have the workspace confirmation modal that has the avatar + name + create button - I wonder if we could just prepend a little warning message into that modal?
![CleanShot 2024-12-06 at 08 58 24@2x](https://github.com/user-attachments/assets/a99b8505-c067-49f0-9b09-7853e7587485)

**Comment by dannymcclain** [comment]
I like Jon's take on the invite pending badge, but agree that we'll need to figure out how that sits with the `Default` badge we're adding to the users default workspace. Actually maybe that's not an issue since your default workspace probably can't be your default until you're fully in it. So probably no issue there.

As for the extra warning, I kinda like Shawn's idea to add a little something extra to the workspace confirmation modal. But if we don't want to do that—what Jon mocked up is what I would expect. If we went that route, I imagine it would flow like this:
1. User clicks new workspace button
2. Dialog Jon mocked up shows (You've already been invited blah blah blah)
3. User clicks New workspace button in the modal
4. We show the standard workspace confirmation dialog

**Comment by dubielzyk-expensify** [comment]
Yeah, fair points all around.

I'm gonna go even further and say that we don't need a warning at all. I kinda think it's weird to warn a user about creating a workspace. With all the new mechanics in place (modal on creation, pending label, tooltips, and this project), I feel like we actually don't need to warn them. If they're creating one on purpose, then the warning is just odd. Also how often do we show it? Only if they have one pending? Every workspace creation? Only for the first one? 

Then secondly, if they create one unknowingly, then I don't think the warning is gonna make a whole lotta sense for them. They still don't know what a workspace is, they just know that one already exists and they might not need another. To me this has signals of something that need solving further up in the flow. But again, I think all the new things coming in actually leaves this warning kinda confusing.

**Comment by shawnborton** [comment]
> With all the new mechanics in place (modal on creation, pending label, tooltips, and this project), I feel like we actually don't need to warn them.

I think that's totally great point. We actually just merged the workspace confirmation modal so I'd be happy to just go with a no warning approach and see how that goes.

**Comment by dannymcclain** [comment]
>With all the new mechanics in place (modal on creation, pending label, tooltips, and this project), I feel like we actually don't need to warn them.

>I'd be happy to just go with a no warning approach and see how that goes

Yeah good points Jon. I'm down with seeing how a no warning approach goes with the new mechanics as well.

**Comment by marcaaron** [comment]
Sounds good. Yeah, I think I ultimately agree that we can wait for this to become a problem. And sounds like we will want to have that workspace row show a "Invite pending". 

@allroundexperts do we have enough information in the front end to make that happen in the next PR? I can also create a new issue for that since this one is getting pretty long - let me know your preference.

**Comment by marcaaron** [comment]
Quick update here that there was a misunderstanding WRT the backend code and design that needs to be addressed. 

To do the design as proposed we need `OpenApp` to return whether a user has accessible domain policies. Getting this information can be quite slow in some cases so my first instinct was to move it to the account validation step (assuming incorrectly that we would always be asking the user for the validate code anyway).

I see now that we will only do this when they have accessible domain policies. In the current state merging the PR will make it so we ask all users on a private domain to enter the magic code. Which would be undesirable and doesn't make sense given the proposed design.

The PR review has gone on for some time and communication has been difficult so I am proposing that we:

- [ ] Push to merge the PR, but not enable the feature for now. We will hardcode something so that this flow does not get engaged yet.
- [ ] Investigate returning something like `hasAccessibleDomainPolicies` in the `OpenApp` call (and take care not to hurt the performance of this command as it is critically and notoriously slow and getting accessible domain policies has the same problems in it's current state)
- [ ] Then we can create a final PR to check `hasAccessibleDomainPolicies` instead of looking at whether the account is on a private domain.

**Comment by marcaaron** [comment]
Working on the final backend parts we need to complete this one.

**Comment by anmurali** [comment]
There are two out of scope features mentioned [here](https://github.com/Expensify/App/issues/48189#issuecomment-2521749184) that will not be a part of this PR, which I agree with. I think we should create separate GHs for each of those and track them. @marcaaron do you agree? If yes - I can create those as separate GHs.

**Comment by marcaaron** [comment]
Yep, sounds good!

**Comment by allroundexperts** [comment]
> @allroundexperts How soon can we get the next PR? Are we waiting on the backend?

Waiting for the issues to be created.

**Comment by allroundexperts** [comment]
> I noticed a issue on main, the onboarding modal does not move to next step from personal details step when login with a private domain user. I believe it is temporary as we are going to create a new PR soon. Posted for tracking...

Fixed it [here](https://github.com/Expensify/App/pull/54364)

**Comment by marcaaron** [comment]
@allroundexperts Still in review, but ready soon!

**Comment by marcaaron** [comment]
@allroundexperts the changes are on production now!

**Comment by marcaaron** [comment]
We can create one more PR and then this should be ready to close once deployed. I am going OOO for some time so we will likely need an internal engineer to take this one over.

**Comment by marcaaron** [comment]
@isabelastisser I think we ignore the above comment until the final PR is merged. So, taking this off HOLD.

**Comment by anmurali** [comment]
> Then we can create a final PR to check `hasAccessibleDomainPolicies` instead of looking at whether the account is on a private domain.

Is this the final PR that @allroundexperts is working on?

**Comment by allroundexperts** [comment]
> @allroundexperts Please ping me when PR is ready. Thanks.

It will be ready on the weekend. I am out of office today and tomorrow.

**Comment by anmurali** [comment]
Do we have an update on this PR @allroundexperts ?

**Comment by allroundexperts** [comment]
Apologies for the delay here. I was a bit under the weather for the past 3 days. I'll try my best to raise a PR today.

**Comment by allroundexperts** [comment]
Blocked because the method to join a workspace that does not have pre-approval fails. 

Asked for help [here](https://expensify.slack.com/archives/C02NK2DQWUX/p1736285798785069)

**Comment by allroundexperts** [comment]
Okay. Seems like I am supposed to use `RequestPolicyAccess` command for joining a workspace that requires approval. However, I don't have the params it requires. @marcaaron any guidance?

**Comment by allroundexperts** [comment]
Here's the [draft](https://github.com/Expensify/App/pull/54912) PR in case someone wants to take a look.

**Comment by isabelastisser** [comment]
@parasharrajat, I just realized that @marcaaron is on leave.

**Comment by isabelastisser** [comment]
asking in convert for reassignment: https://expensify.slack.com/archives/C07HPDRELLD/p1736978559913279

**Comment by anmurali** [comment]
@allroundexperts - where did we land on [this question](https://expensify.slack.com/archives/C07HPDRELLD/p1737403318296179?thread_ts=1736817885.544729&cid=C07HPDRELLD)?

Going to assign you @flodnv so we can see this last bit through.

**Comment by flodnv** [comment]
We are discussing here: https://expensify.slack.com/archives/C07HPDRELLD/p1737992562644379?thread_ts=1736817885.544729&cid=C07HPDRELLD

**Comment by isabelastisser** [comment]
@allroundexperts, can you please provide an update? Thanks!

**Comment by allroundexperts** [comment]
@flodnv Are your changes deployed?

**Comment by flodnv** [comment]
As of 1 hour ago, yes they're in production!

**Comment by flodnv** [comment]
I am done making the API that was blocking the issue. I don't know what else remains, @allroundexperts ?

**Comment by allroundexperts** [comment]
It's on me now. I'll try to add screen recordings tomorrow.

**Comment by allroundexperts** [comment]
Here's the draft [PR](https://github.com/Expensify/App/pull/54912) @parasharrajat

**Comment by danielrvidal** [comment]
@allroundexperts do you have everything you need now to get this one done?

**Comment by allroundexperts** [comment]
@danielrvidal Is https://github.com/Expensify/App/issues/48189#issuecomment-2698700852 merged?

**Comment by flodnv** [comment]
One internal PR was deployed, and the other was reverted due to breaking OpenApp for Expensify employees. I'll have to get back to it this week.

**Comment by blimpich** [comment]
Adding myself on as the CME [as per](https://expensify.slack.com/archives/C07HPDRELLD/p1745204469423139). @parasharrajat @allroundexperts if you need backend changes let me know and I can work with you to help get [the PR out](https://github.com/Expensify/App/pull/54912).

**Comment by blimpich** [comment]
It's merged!

**Comment by blimpich** [comment]
We ended up having to disable the feature before it went out to users. It caused a few deploy blockers. Some were small and easily fixable, but [this deploy blocker](https://github.com/Expensify/App/issues/60828) stumped us.

**Comment by blimpich** [comment]
We'll need to create a PR to re-enable it and also be careful not to change the behavior of other onboarding features like the private domain check.

**Comment by blimpich** [comment]
So @allroundexperts I think the next PR needs to do these things
- re-enable the workspace surfacing feature
- confirm that validate codes are sent on Hybrid App Android/iOS
- confirm that if you sign up with a public domain (example@gmail.com) you are first presented with the "Add Work Email" onboarding screen. I think we broke this, as per [this comment](https://github.com/Expensify/App/issues/60763#issuecomment-2829306726)

Do you think you have bandwidth to do this? All good if not, I know you have a couple projects you are working on right now 👍

**Comment by allroundexperts** [comment]
Hi. I will raise a PR today.

**Comment by allroundexperts** [comment]
PR: https://github.com/Expensify/App/pull/60990

**Comment by blimpich** [comment]
Currently holding on this issue: https://github.com/Expensify/App/issues/61629

Have to fix that first before merging this in.

**Comment by blimpich** [comment]
We merged [the PR](https://github.com/Expensify/App/pull/60990)!

**Comment by blimpich** [comment]
@allroundexperts are these the next steps?

- [ ] wait for the PR to be deployed to staging, QA'ed, and then deployed to prod
- [ ] remove the conditional for hiding this behind a beta (`canUsePrivateDomainOnboarding`)?

**Comment by allroundexperts** [comment]
> [@allroundexperts](https://github.com/allroundexperts) are these the next steps?
> 
> * [ ]  wait for the PR to be deployed to staging, QA'ed, and then deployed to prod[ ]  remove the conditional for hiding this behind a beta (`canUsePrivateDomainOnboarding`)?

Spot on!

**Comment by danielrvidal** [comment]
It looks like [this got deployed](https://github.com/Expensify/App/pull/60990) to production earlier today (I think?) so I think we're good!

**Comment by blimpich** [comment]
Hmm, weird, those [messages at the end of the PR](https://github.com/Expensify/App/pull/60990#issuecomment-2898868484) I think are misleading. They say that I deployed the PR but I didn't, I deployed web. I also think the App deploy has been blocked all day on issues with 1password, among other things. I think we're still waiting on this to get deployed to staging.

**Comment by isabelastisser** [comment]
Payment summary:

@parasharrajat requires payment through NewDot Manual Requests $250 for C+ review
@allroundexperts requires payment through NewDot Manual Requests $250 for proposal

**Comment by isabelastisser** [comment]
@parasharrajat, please complete the BZ checklist before we can process the payment. Thanks!

**Comment by allroundexperts** [comment]
Hey @blimpich!

Can we please discuss the compensation for this project?

In total, we did 6 PR's till now.

https://github.com/Expensify/App/pull/51681
https://github.com/Expensify/App/pull/54364
https://github.com/Expensify/App/pull/54912
https://github.com/Expensify/App/pull/60856
https://github.com/Expensify/App/pull/60990
https://github.com/Expensify/App/pull/62489

Some of these were very large PRs (eg ~41 file changes [here](https://github.com/Expensify/App/pull/51681) 🤯) , and included around 12 issues (bugs included).

That being said, this also required a lot of back and forth backend changes. I started working on this on [Oct 29](https://github.com/Expensify/App/issues/48189#issuecomment-2444852097) and it's Jun now 😄.

I would propose $4000 for the work here. Let me know if it makes sense to you guys.

**Comment by blimpich** [comment]
Hello @allroundexperts! The team discussed the payment for this issue and want to increase the bounty to $3000. We feel that this is a fair offer, given the amount of work. We agree that you and the reviewer put in a lot of time and effort getting this feature across the finish line. However, there were also deploy blockers and regressions that required internal engineering resources, which costs the company money, so we don't think more than $3000 makes sense.

Does that sound agreeable to you? If so I'll change the price and get the process rolling to pay this out.

**Comment by allroundexperts** [comment]
Sounds good. Thank you @blimpich. Let's proceed.

**Comment by parasharrajat** [comment]
Sounds good to me as well.

**Comment by parasharrajat** [comment]
## Regression Test Proposal

### Test:


#### When there are no existing workspaces from the same domain. 

1. Log in with a new private domain account that has never been used on Expensify. 
2. You should first see the onboarding Purpose step. 
3. Click continue, and onboarding should happen as normal for all users. 

#### When there are existing workspaces from the same domain. 

1. Log in with a new private domain account.
2. Verify that an onboarding modal appears, you should see the Personal details onboarding step.
3. Enter your name and click continue, the user should be taken to the validate email step to enter OTP. 
4. Enter the OTP, and verify that after clicking next, you're shown a list of the existing workspaces to join.
 
   ##### Pre-approval not required:
       1. Join any workspace by clicking the `join now` button on any workspace, and verify that you see the new workspace in the LHN. 
 
   ##### Pre-approval is required:
       - Make sure that another user exists on the same private domain and has a workspace created that has pre-approval of new members turned off. This needs to be done through Old dot, by going over to the workspace settings, and navigating to the members section.


        1. Verify that the list contains the pre-approval workspace, and the pre-approval workspace has a button that reads `Ask to join`.
        5. Click on the button.
        6. Verify that you have landed back on LHN.
        7. Log in as an admin of the selected workspace and go to the #admins chat, and make sure that you see a message asking that the new user intends to join the workspace.
        8. Accept the request.
        9. Log back in with the previous email from step 1 and verify that you see the workspace chat in LHN.

Do we agree 👍 or 👎

**Comment by isabelastisser** [comment]
Payment summary:

Reviewer: @parasharrajat owed $3000 via NewDot for C+ review
Reviewer: @allroundexperts owed $3000 via NewDot for proposal

**Comment by parasharrajat** [comment]
@isabelastisser We can close this issue as well if nothing is pending here. Payment can be requested later. Thanks.

**Comment by isabelastisser** [comment]
All set now.

**Comment by JmillsExpensify** [comment]
$3,000 approved for @allroundexperts
